### PR TITLE
Add networking benchmarks for libp2p

### PR DESCRIPTION
**Pull request by AndreiEres**
# Description

Implemented benchmarks for Notifications and RequestResponse protocols with libp2p implementation. These benchmarks allow us to monitor regressions and implement fixes before they are observed in real chain. In the future, they can be used for targeted optimizations of litep2p compared to libp2p.

Part of https://github.com/paritytech/polkadot-sdk/issues/5220

Next steps:
- Add benchmarks for litep2p implementation
- Optimize load to get better results
- Add benchmarks to CI to catch regressions

## Integration

Benchmarks don't affect downstream projects.

### CODE REVIEW

File: substrate/client/network/benches/request_response_protocol.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+		metrics_registry: None,
+		bitswap_config: None,
+		notification_metrics: NotificationMetrics::new(None),
+	})
+	.unwrap();
+
+	(worker, rx, notification_service)
+}
+
+async fn run_consistently(size: usize, limit: usize) {
```
Comment (dmitry-markin) [inline substrate/client/network/benches/request_response_protocol.rs:108]: Nit: `run_serially`?

File: substrate/client/network/benches/request_response_protocol.rs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	let network1_run = worker1.run();
+	let network2_run = worker2.run();
+	tokio::pin!(requests);
+	tokio::pin!(network1_run);
+	tokio::pin!(network2_run);
+
+	loop {
+		tokio::select! {
+			_ = &mut network1_run => {},
+			_ = &mut network2_run => {},
```
Comment (dmitry-markin) [inline substrate/client/network/benches/request_response_protocol.rs:143]: The same as with notifications, we better spawn a separate tokio task for every "node".

File: substrate/client/network/benches/request_response_protocol.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+					result: Ok(vec![0; size]),
+					reputation_changes: vec![],
+					sent_feedback: None,
+				}).unwrap();
+			}
+		}
+	}
+}
+
+async fn run_with_backpressure(size: usize, limit: usize) {
```
Comment (dmitry-markin) [inline substrate/client/network/benches/request_response_protocol.rs:157]: I have checked the [libp2p request-response implementation](https://github.com/libp2p/rust-libp2p/blob/812a7cdf5d229610bb8eda1f106eee5ac3f699d2/protocols/request-response/src/lib.rs#L425), and it does not provide any backpressure feedback. So this benchmark is useless until we implement it for litep2p.

**Review by dmitry-markin — APPROVED** [review]
Looks pretty good!

**Review by lexnv — APPROVED** [review]
Nice job here! 🙏

**Pull request merged**